Add JUnitXml to OutputFormat

bgelens opened the pull request:
<!--

Thank you for contributing to Pester! Please provide a descriptive title of the pull request in the field 'Title'.

-->

## 1. General summary of the pull request

Fixes #1346 

This PR will be in draft so progress on the JUnitXml implementation can be tracked / discussed. Once done, it will be promoted to a generic PR

<!--

Please describe what your pull request fixes, or how it improves Pester.

If your pull request resolves a reported issue, please mention it by using  `Fix #<issue_number>` on a new line, this will close the linked issue automatically when this PR is merged. For more info see: [Closing issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/).

If your pull request integrates Pester with another system, please tell us how the change can be tested.

Please remember to update [the Pester wiki](https://github.com/pester/Pester/wiki) if needed.

Before you continue, please review [Contributing to Pester](https://github.com/pester/Pester/wiki/Contributing-to-Pester) and [Development rules - technical](https://github.com/pester/Pester/wiki/Developement-rules---technical).

Our continuous integration system doesn't send any notifications about failed tests. Please return to the opened pull request (after ~60 minutes) to check if is everything OK.

-->

bgelens commented:
@nohwnd The reports generated with the current code changes are correctly interpreted in GitLab (I've seen AzDevOps also supports JUnit so I'm going to test there a bit as well).

Could you do an initial review pass and tell me what you are not happy with :) I've copied the NUnit routine as much as possible.

nohwnd commented on Functions/TestResults.ps1:
```diff
@@ -33,23 +33,26 @@ function Export-PesterResults {
-        [String]$Path
+        [String]$Path,
+
+        [parameter(Mandatory = $true)]
```
Would be nice to use `ValidateRange` with the supported formats.

nohwnd commented on Functions/TestResults.ps1:
```diff
@@ -146,7 +157,7 @@ function Write-NUnitEnvironmentInformation($PesterState, [System.Xml.XmlWriter]
 
     $environment = Get-RunTimeEnvironment
-    foreach ($keyValuePair in $environment.GetEnumerator()) {
+    foreach ($keyValuePair in $environment.GetEnumerator() | Where-Object -FilterScript {$_.Name -ne 'junit-version'}) {
```
This might fail when user mocks `Where-Object`. It's not a big issue here, because at this point all Mocks should be removed, but pipeline is slow, and keeping track of where we can and cannot use cmdlets is difficult, so new code should rather not use cmdlets :) use `if  {$_.Name -eq 'junit-version'} { continue }` instead please. If there is no other option and you need to use cmdlet please use it via SafeCommands hashtable.

bgelens commented:
I've addressed your comments.

I think what is needed now before we consider to move this forward is:

* [ ] Add tests
* [ ] Test on other CI systems than GitLab
  * [ ] Azure DevOps
  * [ ] ...?

renehernandez commented on Functions/TestResults.ps1:
```diff
@@ -222,6 +238,131 @@ function Write-NUnitTestSuiteElements($Node, [System.Xml.XmlWriter] $XmlWriter,
+
+    Write-JUnitTestSuiteAttributes -Action $Node -XmlWriter $XmlWriter -Package $Package -Id $Id
+
+    if ($Node.Actions.Type -eq 'TestCase') {
```
You could refactor this as follows:

```
$actions = $Node.Actions.Actions

if ($Node.Actions.Type -eq 'TestCase') {
    $actions = $Node.Actions
}

foreach ($t in $actions) {
    Write-JUnitTestCaseElements -Action $t -XmlWriter $XmlWriter
}
```

That way you reduce code duplication

renehernandez commented on Functions/TestResults.ps1:
```diff
@@ -222,6 +238,131 @@ function Write-NUnitTestSuiteElements($Node, [System.Xml.XmlWriter] $XmlWriter,
+        }
+    }
+
+    $XmlWriter.WriteAttributeString('status', $Action.Result)
```
Should not this use `$status` instead of `$Action.Result`?

bgelens replied:
I've updated the code a bit to make is more clear what it is used for. 

I'm using the original status for the attribute as its value is not predefined in the schema (keeping terms like "Inconclusive" as part of the junit report). Next, a skipped or failure type adds extra elements which value is predefined in the schema and widgets like the one from GitLab explicitly search for)

bgelens commented:
Azure Pipelines result:

![image](https://user-images.githubusercontent.com/10071039/63212539-5f0a8400-c106-11e9-94f0-aacc1156faf6.png)

![image](https://user-images.githubusercontent.com/10071039/63212548-8cefc880-c106-11e9-963d-bcdef2a83f56.png)

I think it is looking good :)

bgelens commented:
Adding the test file name to class= will make Azure DevOps aware (filter allows to check the box of the .tests.ps1 file) of where the test originated from so I'll update this. I'll also will create the same testname as with NUnitXml `Describe.Context.TestName`

renehernandez commented:
@bgelens It looks good

bgelens commented:
Generating junit based on these tests: [simple.tests.ps1.txt](https://github.com/pester/Pester/files/3512290/simple.tests.ps1.txt)

Result in this xml: [junit.xml.txt](https://github.com/pester/Pester/files/3512291/junit.xml.txt)

In Azure DevOps this now looks like:

![image](https://user-images.githubusercontent.com/10071039/63215769-c5ee6400-c12b-11e9-89b1-b563a39a22bf.png)

Where the name of the test is the path Describe.Context.It or Describe.It / Context.It or It

I think it is good enough unless someone with actual JUnit experience has some comments :)

I'll work on some tests in the next few days and will remove the draft once that is done

bgelens commented:
@nohwnd you offered to help with the tests right :)

I started adding test by duplicating and adjusting what was already there for nunit.

I apparently don't know enough about PesterState to imagine how I would get PSv2 tests to pass. I get that a property is missing but afaict, `Enter-TestGroup` calls `New-TestGroup` which should add the `TotalCount` property.

I need to imagine for PSv2 as I'm working on Mac with PSv7.

Can you take a look? Help out :)

bgelens commented:
hey @nohwnd, Any idea why the CI is not triggered?

bgelens commented:
Turns out I had a merge conflict as master was changed but this wasn't obviously presented.
I rebased and apparently still need to fix the psv2 issues :|

bgelens commented:
I think I fixed the issues by handling things a little bit differently. I'm going to do some validation with reports. If it looks ok, I'm going to remove the draft status

bgelens commented:
I've validated the results to be correct in both GitLab and Azure DevOps. Ready for review!

nohwnd commented:
@bgelens I know I said I will help with the tests, but I did not and I am sorry. But on the other hand you did it all by yourself without anyone "holding your hand".  🥳🕺🍰 It's merged and will be released soon :)
